giraffe/src/transforms/mosaic.ts

```diff
@@ -1,121 +1,180 @@
 import {newTable} from '../utils/newTable'
 import {MosaicLayerSpec, Table} from '../types'
-import {X_MIN, X_MAX, FILL, SERIES} from '../constants/columnKeys'
+import {DISPLAY_NAME, FILL, SERIES, X_MAX, X_MIN} from '../constants/columnKeys'
 import {createGroupIDColumn} from './'
 import {resolveDomain} from '../utils/resolveDomain'
 import {getNominalColorScale} from './'
 
 export const mosaicTransform = (
   inputTable: Table,
   xColumnKey: string,
-  yColumnKey: string,
+  yColumnKeys: string[],
+  yLabelColumns: string[],
+  yLabelColumnSeparator: string,
   xDomain: number[],
   fillColKeys: string[],
   colors: string[]
 ): MosaicLayerSpec => {
-  const xInputCol = inputTable.getColumn(xColumnKey, 'number')
-  const yInputCol = inputTable.getColumn(yColumnKey, 'string')
   const [fillColumn, fillColumnMap] = createGroupIDColumn(
     inputTable,
     fillColKeys
   )
+  const xInputCol = inputTable.getColumn(xColumnKey, 'number') || []
+
+  let labelColumns = Array.isArray(yLabelColumns)
+    ? yLabelColumns.filter(labelColumn => yColumnKeys.includes(labelColumn))
+    : []
+  if (labelColumns.length === 0 && Array.isArray(yColumnKeys)) {
+    labelColumns = yColumnKeys
+  }
+
+  const yInputCols = {}
+  if (Array.isArray(yColumnKeys)) {
+    yColumnKeys.forEach(columnKey => {
+      const column = inputTable.getColumn(columnKey, 'string')
+      yInputCols[columnKey] = column
+    })
+  }
+
+  // Mosaic can only have one column as the fill value,
+  //   always the first fill column key
   const valueKey = fillColumnMap.columnKeys[0]
 
+  const timeStampMap = {}
+
+  for (let i = 0; i < inputTable.length; i++) {
+    const yColumnTick = Array.isArray(yColumnKeys)
+      ? yColumnKeys.reduce((combinedValue, key) => {
```

Comment:
No very big. However, this is a good point. Will look into improving the complexity. It might not be possible but this is definitely due another round of scrutiny for refactoring. Will take a look.

Reply:
After further analysis, there isn't a quicker way.

Imagine a two-dimensional grid, like a giant field of crops uniformly spaced in perfect rows. A farmer wants to go through each row and harvest a certain number of plants. The plants being harvested are at the same position in each row. No matter how the farmer does it, he has to do the work to visit each plant being harvested.

Similarly, we are doing that here.

You will notice a second similar **.reduce** call which is being done on another array that is a subset of the first array. This second array is entirely based on user selection and there is no predicting how long it is other than it must be a subset of the first array. In our analogy, this would be the farmer doing a second harvest but smaller than the first.

Reply:
great explanation! keep that farmer metaphor in your pocket for explaining this to people :)